File: diamond/elixir/diamond/lib/diamond.ex
defmodule Diamond do
	def letters(char) do
			"ABCDEFGHIJKLMNOPQRSTUVWXYZ"
		|> String.codepoints
		|> Enum.take_while &(&1 <= char)
	end

	def spaces_after(chars_count) do
		0..chars_count - 1
		|> Enum.map &(String.duplicate " ", &1)
	end

	def spaces_before(chars_count) do
		spaces_after(chars_count)
		|> Enum.reverse
	end

	def lines(char) do
		chars = letters char
		spaces = Enum.count(chars)
		for {front, letter, back} <-
		List.zip(
			[
				spaces_before(spaces),
				chars,
				spaces_after(spaces)
			]),
		do: front <> letter <> back
	end

	def mirror_vertical(lines) do
		reversed = lines
		|> Enum.map(&(String.reverse(&1)))
		|> Enum.map(&(String.slice(&1, 1..-1)))
		
		for {a, b} <- Enum.zip(lines, reversed), do: a <> b
	end

	def mirror_horizontal(lines) do
		lines ++ tl(Enum.reverse(lines))
	end

	def all_lines(char) do
		lines(char)
		|> mirror_vertical
		|> mirror_horizontal
	end

	def print(char) do
		Enum.join(all_lines(char), "\n") <> "\n"
	end
end

File: diamond/c++/v1/diamond.cpp
#include "diamond.hpp"
#include <boost/range/irange.hpp>
#include <boost/range/algorithm.hpp>
#include <boost/range/numeric.hpp>
#include <boost/range/combine.hpp>
#include <iterator>
#include <string>
#include <sstream>

namespace v1
{
  using namespace std::literals::string_literals;
  
  std::string Diamond::letters(int amount)
  {
    return "ABCDEFGHIJKLMNOPQRSTUVWXYZ"s.substr(0, amount);
  }

  Diamond::Lines Diamond::spaces_after(int amount)
  {
    auto spaces = Lines(amount);
    boost::transform(boost::irange(0, amount),
		     begin(spaces),
		     [](const auto& num_of_spaces)
		     { return std::string(num_of_spaces, ' '); });

    return spaces;
  }

  Diamond::Lines Diamond::spaces_before(int amount)
  {
    auto spaces = spaces_after(amount);
    return boost::reverse(spaces);
  }
  
  Diamond::Lines Diamond::lines(int amount)
  {
    Lines lines(amount);
    boost::transform(boost::combine(spaces_before(amount),
				    letters(amount),
				    spaces_after(amount)),
		     begin(lines),
		     [](const auto& space_letter_space)
		     {
		       std::string before, letter, after;
		       boost::tie(before, letter, after) = space_letter_space;
		       return before + letter + after;
		     });
    return lines;
  }
  
  Diamond::Lines Diamond::mirror_vertical(Lines lines)
  {
    for(auto& line : lines) line += std::string(crbegin(line) + 1, crend(line));
    return lines;
  }

  Diamond::Lines Diamond::mirror_horizontal(Lines lines)
  {
    lines.reserve(2 * lines.size() - 1);
    copy(++crbegin(lines), crend(lines),
	 back_inserter(lines));
    return lines;
  }

  Diamond::Lines Diamond::all_lines(int amount)
  {
    return mirror_horizontal(mirror_vertical(lines(amount)));
  }

  int Diamond::char_to_num(char c)
  {
    return c - 'A' + 1;
  }

  std::string Diamond::print(char c)
  {
    const auto lines = all_lines(char_to_num(c));
    return boost::accumulate(lines, ""s,
			     [](const auto& diamond, const auto& line)
			     {
			       return diamond + line + '\n';
			     });
  }
}
